Deprecate RTTI-specific functions in stream's Format

mathias-lang-sociomantic opened the pull request:
Deprecate a few methods as part of https://github.com/sociomantic-tsunami/ocean/pull/361

mathias-lang-sociomantic committed:
Deprecation Arguments.error method

It's current signature is tighly tied to RTTI and Tango Layout, and cannot be converted to be Formatter-compatible.

mathias-lang-sociomantic committed:
Deprecate RTTI-specific functions in stream's Format

We cannot simply turn 'format' from RT args to CT args in v4.0.0, because that would try to format the TypeInfo[] and void*[],
and thus might introduce hard to spot bugs. Instead, with this extra deprecation, we'll catch any remaining user of the
functionality.

mathias-lang-sociomantic committed:
Remove unused debug blocks

leandro-lucarella-sociomantic commented on src/ocean/text/Arguments.d:
```diff
@@ -611,6 +611,7 @@ import ocean.io.stream.Format : FormatOutput;
 import ocean.math.Math;
 import ocean.text.Util;
 import ocean.text.convert.Formatter;
+static import ocean.text.convert.Layout_tango; // TODO: Remove in v4.0.0
```
Create an issue for this?

mathias-lang-sociomantic replied:
https://github.com/sociomantic-tsunami/ocean/issues/206

gavin-norman-sociomantic replied:
You can also include the keyword "deprecated" in the comment, so we'll find it when searching for old crap to remove.

leandro-lucarella-sociomantic commented on src/ocean/text/Arguments.d:
```diff
@@ -1096,6 +1097,7 @@ public class Arguments
 
     ***************************************************************************/
 
+    deprecated("Use the overload that doesn't take a parameter (and use the Formatter")
```
Missing an ending `)`

leandro-lucarella-sociomantic commented on src/ocean/text/Arguments.d:
```diff
@@ -1114,6 +1116,26 @@ public class Arguments
+        {
+            if (arg.error)
+            {
+                // TODO: Replace with `sformat` in v4.0.0
```
Also issue

mathias-lang-sociomantic replied:
Also covered by https://github.com/sociomantic-tsunami/ocean/issues/206

leandro-lucarella-sociomantic commented on src/ocean/io/Stdout.d:
```diff
@@ -148,7 +149,7 @@ public class TerminalOutput ( T ) : FormatOutput!(T)
     public typeof(this) format ( Const!(T)[] fmt, ... )
     {
-        super.format(fmt, _arguments, _argptr);
+        super.doNotUse(fmt, _arguments, _argptr);
```
Say what? :thinking:

leandro-lucarella-sociomantic commented on src/ocean/io/stream/Format.d:
```diff
@@ -138,7 +127,14 @@ class FormatOutput(T) : OutputFilter
+        }
+
+        /// Used by derived classes to avoid deprecations - Remove in v4.0.0
+        final void doNotUse (Const!(T)[] fmt, TypeInfo[] arguments, ArgList args)
```
Also, create an issue not to forget and maybe give it a name that says what it does (besides making it obvious it shouldn't be used).

mathias-lang-sociomantic replied:
I already raised a P.R. against v4.x.x to remove it. Also the comment says why it's called this way. However, it should be `protected`

leandro-lucarella-sociomantic reviewed (commented):
The last commit message says: `* submodules/makd v1.9.1(8711da2)...v1.6.0(949d0ad) (50 commits)`. I guess it's wrong, right?

mathias-lang-sociomantic commented:
Updated:
- Removed wrong Makd downgrade
- Added missing `)`
- Added `protected` on `doNotUse`.

leandro-lucarella-sociomantic commented:
> Added protected on doNotUse.

Still not crazy about this (I meant name it something like `_transitional_format` wouldn't be enough / better?)

mathias-lang-sociomantic commented:
Well I'm not set on the name. I just consider it more important to say "Do not use this" than to say what it's for... Because it's not meant to be used, so its purpose had no value for external users.
But as I said, I am not going to argue about it, so I changed the name to `_transitional_format`.

leandro-lucarella-sociomantic commented:
> Well I'm not set on the name. I just consider it more important to say "Do not use this" than to say what it's for... Because it's not meant to be used, so its purpose had no value for external users.

But for people that need to read/maintain the code (not library users but library developers) is still important to be able to understand the code :)

gavin-norman-sociomantic commented on src/ocean/text/Arguments.d:
```diff
@@ -1114,6 +1116,26 @@ public class Arguments
         return result;
     }
 
+    /// Ditto
```
Not ditto, as the params are different.

mathias-lang-sociomantic replied:
Then by definition you can never use `ditto` for the same overload set.
AFAIK ddoc supports that usage (and it's widely used in Phobos), but maybe we should document in which case we can use ditto then, since expectations seems to vary.

gavin-norman-sociomantic replied:
Seems so. My expectation is that literally the same comment would be applicable to the `ditto`ed thing.

mathias-lang-sociomantic replied:
Well actually most the the description was wrong as well, so Ditto was not applicable here.

gavin-norman-sociomantic commented on src/ocean/io/Stdout.d:
```diff
@@ -134,6 +134,7 @@ public class TerminalOutput ( T ) : FormatOutput!(T)
 
     ***************************************************************************/
 
+    deprecated("Using Layout with this class is deprecated")
```
Suggested course of action?

gavin-norman-sociomantic commented on src/ocean/io/stream/Format.d:
```diff
@@ -120,6 +108,7 @@ class FormatOutput(T) : OutputFilter
 
         **********************************************************************/
 
+        deprecated("Using Layout with this class is deprecated")
```
Suggested course of action?

mathias-lang-sociomantic commented:
Fixed according to @gavin-norman-sociomantic 's feedback.

mihails-strasuns-sociomantic commented on src/ocean/io/Stdout.d:
```diff
@@ -59,10 +59,10 @@ static this ( )
 {
     // note that a static-ctor inside Layout fails
     // to be invoked before this is executed (bug)
```
is comment still applicable? i would expect not, if new code works

mathias-lang-sociomantic replied:
The only static ctor I could find is here: https://github.com/sociomantic-tsunami/ocean/blob/edf92dc7eea6869b8e54d99fe11f9a452a8ab1fe/src/ocean/text/convert/Layout_tango.d#L1033-L1036
I'll remove the comment

mihails-strasuns-sociomantic commented on src/ocean/io/stream/Format.d:
```diff
@@ -138,7 +127,14 @@ class FormatOutput(T) : OutputFilter
+            return this;
+        }
+
+        /// Used by derived classes to avoid deprecations - Remove in v4.0.0
```
Please mention "deprecated" substring here too.

mihails-strasuns-sociomantic reviewed (commented):
Few minor comments

mathias-lang-sociomantic committed:
Remove outdated comment

The only static dtor inside Layout is to instantiate an instance for unittests.
There is nothing that could affect this code. The ctor mentioned was most likely moved to the Format module at some point.

mathias-lang-sociomantic commented:
Updated

mihails-strasuns-sociomantic reviewed (approved):
LGTM, @gavin-norman-sociomantic please merge if you have no further requests
